corehq/apps/reports/standard/users/reports.py

```diff
@@ -0,0 +1,133 @@
+
+    fields = [
+        'corehq.apps.reports.filters.users.ExpandedMobileWorkerFilter',
+        'corehq.apps.reports.filters.users.ChangedByUserFilter',
+        'corehq.apps.reports.filters.users.ChangeActionFilter',
+        'corehq.apps.reports.filters.dates.DatespanFilter',
+    ]
+
+    description = ugettext_lazy("History of user updates")
+    ajax_pagination = True
+
+    sortable = False
+
+    @property
+    def default_datespan(self):
+        return datespan_from_beginning(self.domain_object, self.timezone)
+
+    @property
+    def headers(self):
+        h = [
+            DataTablesColumn(_("User")),
+            DataTablesColumn(_("By User")),
+            DataTablesColumn(_("Action")),
+            DataTablesColumn(_("Via")),
+            DataTablesColumn(_("Change Message")),
+            DataTablesColumn(_("Changes")),
+            DataTablesColumn(_("Timestamp")),
+        ]
+
+        return DataTablesHeader(*h)
+
+    @property
+    def total_records(self):
+        return self._get_queryset().count()
+
+    @memoized
+    def _get_queryset(self):
+        user_ids = self._get_user_ids()
+        by_user_ids = self._get_by_user_ids()
+        actions = self.request.GET.getlist('action')
+        query = self._build_query(user_ids, by_user_ids, actions)
+        return query
+
+    def _get_user_ids(self):
+        es_query = self._get_users_es_query(self.request.GET.getlist(EMWF.slug))
+        return es_query.values_list('_id', flat=True)
+
+    def _get_by_user_ids(self):
+        es_query = self._get_users_es_query(self.request.GET.getlist(ChangedByUserFilter.slug))
+        return es_query.values_list('_id', flat=True)
+
+    def _get_users_es_query(self, slugs):
+        return EMWF.user_es_query(
+            self.domain,
+            slugs,
+            self.request.couch_user,
+        )
+
+    def _build_query(self, user_ids, by_user_ids, actions):
+        filters = Q(domain=self.domain)
+
+        if user_ids:
+            filters = filters & Q(user_id__in=user_ids)
+
+        if by_user_ids:
+            filters = filters & Q(by_user_id__in=by_user_ids)
+
+        if actions and ChangeActionFilter.ALL not in actions:
+            filters = filters & Q(action__in=actions)
+
+        if self.datespan:
+            filters = filters & Q(created_at__lt=self.datespan.enddate_adjusted,
+                                  created_at__gte=self.datespan.startdate)
+        return UserHistory.objects.filter(filters)
+
+    @property
+    def rows(self):
+        records = self._get_queryset().order_by('-created_at')[
+            self.pagination.start:self.pagination.start + self.pagination.count
+        ]
+        for record in records:
+            yield _user_history_row(record)
+
+
+def _user_history_row(record):
+    return [
+        CouchUser.get_by_user_id(record.user_id).username,  # ToDo: re-think this fetch
+        CouchUser.get_by_user_id(record.by_user_id).username,
```

Comment:
Hey @snopoke @orangejenny ,

Do you have any suggestions about this?

A User History record has the couch user id and we need to show the username.
For the new tracking,
I chose to use couch user ID instead of django user ID (which was being used earlier which made it possible to [fetch users with a JOIN](https://github.com/dimagi/commcare-hq/pull/29681/files#diff-97674072217e37379be148b9fdf743f686c009d5d9741d8ca3607fa16a4d526fR79)) because Couch users are "more HQ" and did help with making logging better at places like [this one](https://github.com/dimagi/commcare-hq/pull/29902/files#r650774616).

Though that makes it a bit harder here, to show usernames, in report since there can't be a JOIN. So either,
1. `get_by_user_id` is cached so this is okay to leave it like this and may be additionally, only fetch, if not already fetched.
2. get them in bulk from couch (looks like there is no way to fetch user docs in bulk so would need to add a way). I guess we can't get it from ES because deleted users are not present in ES.
3. add django user ID for both (user and by_user) and use that to JOIN and fetch username.
4. store usernames on the user history record itself as meta info, if not already available and use that for display

Note: we have user's username in the user json stored in the record itself for Create and Delete.

Reply:
My gut is that 1 isn't great, even though user id is cached, since it's still potentially 100 couch queries to show the first page of records (one per row).

> 2. get them in bulk from couch (looks like there is no way to fetch user docs in bulk so would need to add a way)

Doesn't `get_all_user_id_username_pairs_by_domain` do that?

This seems like a good option, and better than storing redundant data on the user history record.

Reply:
...nevermind, `get_all_user_id_username_pairs_by_domain` doesn't do what you want and I don't see django user id stored in any of the `users` couch views. I hate suggesting this, but I wonder if a new couch view that includes django user id is the way to go.

Reply:
You can use `corehq.apps.users.util.cached_user_id_to_username` which is used by exports and a few other places.